components/nifosg/controller.cpp

```diff
@@ -309,6 +309,50 @@ void VisController::operator() (osg::Node* node, osg::NodeVisitor* nv)
     traverse(node, nv);
 }
 
+RollController::RollController(const Nif::NiFloatData *data)
+    : mData(data->mKeyList, 1.f)
+{
+}
+
+RollController::RollController()
+{
+}
+
+RollController::RollController(const RollController &copy, const osg::CopyOp &copyop)
+    : osg::NodeCallback(copy, copyop)
+    , Controller(copy)
+    , mData(copy.mData)
+    , mStartingTime(0)
+{
+}
+
+void RollController::operator() (osg::Node* node, osg::NodeVisitor* nv)
+{
+    traverse(node, nv);
+
+    if (hasInput())
+    {
+        double newTime = nv->getFrameStamp()->getSimulationTime();
+        double duration = newTime - mStartingTime;
+        mStartingTime = newTime;
+        if (duration == 0)
+            return;
+
+        float value = mData.interpKey(getInputValue(nv));
+        osg::MatrixTransform* transform = static_cast<osg::MatrixTransform*>(node);
+        osg::Matrix matrix = transform->getMatrix();
+
+        // Rotate around "roll" axis.
+        // Note: we set per-update speed, not per-second, so an actual rotation speed will depend on FPS
+        // Original engine behaves in the similar way, but actual rotation speed can be different from OpenMW.
+        osg::Quat rotate (value, osg::Vec3f(0,0,1));
+        osg::Quat orient = rotate * matrix.getRotate();
+        matrix.setRotate(orient);
```

Comment:
To close 4856: bzzt believes the external rotation of the node must be accounted for in orient (which makes sense, though I'm not sure what else could rotate the thing).

Reply:
... what?